File: example/image-classification/train_imagenet.py
import find_mxnet
import mxnet as mx
import logging
import argparse
import os
import train_model

# don't use -n and -s, which are resevered for the distributed training
parser = argparse.ArgumentParser(description='train an image classifer on imagenet')
mutually_exclusive_parser_group = parser.add_mutually_exclusive_group(required=True)
parser.add_argument('--network', type=str, default='inception-bn',
                    choices = ['alexnet', 'vgg', 'googlenet', 'inception-bn',
                               'inception-bn-full', 'inception-v3', 'resnet'],
                    help = 'the cnn to use')
mutually_exclusive_parser_group.add_argument('--data-dir', type=str,
                    help='the input data directory')
parser.add_argument('--model-prefix', type=str,
                    help='the prefix of the model to load')
parser.add_argument('--save-model-prefix', type=str,
                    help='the prefix of the model to save')
parser.add_argument('--lr', type=float, default=.01,
                    help='the initial learning rate')
parser.add_argument('--lr-factor', type=float, default=1,
                    help='times the lr with a factor for every lr-factor-epoch epoch')
parser.add_argument('--lr-factor-epoch', type=float, default=1,
                    help='the number of epoch to factor the lr, could be .5')
parser.add_argument('--clip-gradient', type=float, default=5.,
                    help='clip min/max gradient to prevent extreme value')
parser.add_argument('--num-epochs', type=int, default=20,
                    help='the number of training epochs')
parser.add_argument('--load-epoch', type=int,
                    help="load the model on an epoch using the model-prefix")
parser.add_argument('--batch-size', type=int, default=32,
                    help='the batch size')
parser.add_argument('--gpus', type=str,
                    help='gpus to be used, e.g "0,1,2,3"')
parser.add_argument('--kv-store', type=str, default='local',
                    help='the kvstore type')
parser.add_argument('--num-examples', type=int, default=1281167,
                    help='the number of training examples')
parser.add_argument('--num-classes', type=int, default=1000,
                    help='the number of classes')
parser.add_argument('--log-file', type=str,
		    help='the name of log file')
parser.add_argument('--log-dir', type=str, default="/tmp/",
                    help='directory of the log file')
parser.add_argument('--train-dataset', type=str, default="train.rec",
                    help='train dataset name')
parser.add_argument('--val-dataset', type=str, default="val.rec",
                    help="validation dataset name")
parser.add_argument('--data-shape', type=int, default=224,
                    help='set image\'s shape')
mutually_exclusive_parser_group.add_argument('--benchmark', default=False, action='store_true',
                    help='benchmark for 50 iterations using randomly generated Synthetic data')
args = parser.parse_args()

# network
import importlib
net = importlib.import_module('symbol_' + args.network).get_symbol(args.num_classes)


# data
import random
from mxnet.io import DataBatch, DataIter
import numpy as np
class SyntheticDataIter(DataIter):
    def __init__(self, num_classes, data_shape, max_iter):
        self.batch_size = data_shape[0]
        self.cur_iter = 0
        self.max_iter = max_iter
        label = np.random.randint(0, num_classes, [self.batch_size,])
        data = np.random.uniform(-1, 1, data_shape)
        self.data = mx.nd.array(data)
        self.label = mx.nd.array(label)
    def __iter__(self):
        return self
    @property
    def provide_data(self):
        return [('data',self.data.shape)]
    @property
    def provide_label(self):
        return [('softmax_label',(self.batch_size,))]
    def next(self):
        self.cur_iter += 1
        if self.cur_iter <= self.max_iter:
            return DataBatch(data=(self.data,),
                             label=(self.label,),
                             pad=0,
                             index=None,
                             provide_data=self.provide_data,
                             provide_label=self.provide_label)
        else:
            raise StopIteration
    def __next__(self):
        return self.next()
    def reset(self):
        self.cur_iter = 0

def get_sythentic_data_iter(args, kv):
    data_shape = (args.batch_size, 3, args.data_shape, args.data_shape)
    train = SyntheticDataIter(args.num_classes, data_shape, 50)
    val = SyntheticDataIter(args.num_classes, data_shape, 1)
    return (train, val)

def get_iterator(args, kv):
    data_shape = (3, args.data_shape, args.data_shape)
    train = mx.io.ImageRecordIter(
        path_imgrec = os.path.join(args.data_dir, args.train_dataset),
        mean_r      = 123.68,
        mean_g      = 116.779,
        mean_b      = 103.939,
        data_shape  = data_shape,
        batch_size  = args.batch_size,
        rand_crop   = True,
        rand_mirror = True,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)

    val = mx.io.ImageRecordIter(
        path_imgrec = os.path.join(args.data_dir, args.val_dataset),
        mean_r      = 123.68,
        mean_g      = 116.779,
        mean_b      = 103.939,
        rand_crop   = False,
        rand_mirror = False,
        data_shape  = data_shape,
        batch_size  = args.batch_size,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)

    return (train, val)

if args.benchmark:
    train_model.fit(args, net, get_sythentic_data_iter)
else:
    train_model.fit(args, net, get_iterator)

File: example/image-classification/train_imagenet.R
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# 
# This file shows how to train ImageNet dataset with several Convolutional Neural Network architectures in R.
# More information: https://blogs.technet.microsoft.com/machinelearning/2016/11/15/imagenet-deep-neural-network-training-using-microsoft-r-server-and-azure-gpu-vms/
#
# To train ResNet-18:
# Rscript train_imagenet.R --network resnet --depth 18 --batch-size 512 --lr 0.1 --lr-factor 0.94 --gpu 0,1,2,3 --num-round 120 /
# --data-dir /path/to/data --train-dataset train.rec --val-dataset val.rec --log-dir $PWD --log-file resnet18-log.txt /
# --model-prefix resnet18 --kv-store device
#

# Train imagenet
require(mxnet)
require(argparse)

# Iterator
get_iterator <- function(args) {
  data.shape <- c(args$data_shape, args$data_shape, 3)
  train = mx.io.ImageRecordIter(
    path.imgrec     = file.path(args$data_dir, args$train_dataset),
    batch.size      = args$batch_size,
    data.shape      = data.shape,
    mean.r          = 123.68,
    mean.g          = 116.779,
    mean.b          = 103.939,
    rand.crop       = TRUE,
    rand.mirror     = TRUE
  )
  
  val = mx.io.ImageRecordIter(
    path.imgrec     = file.path(args$data_dir, args$val_dataset),
    batch.size      = args$batch_size,
    data.shape      = data.shape,
    mean.r          = 123.68,
    mean.g          = 116.779,
    mean.b          = 103.939,
    rand.crop       = FALSE,
    rand.mirror     = FALSE
  )
  ret = list(train=train, value=val)
}

# parse arguments
parse_args <- function() {
  parser <- ArgumentParser(description='train an image classifer on ImageNet')
  parser$add_argument('--network', type='character', default='resnet',
                      choices = c('resnet', 'inception-bn', 'googlenet', 'inception-resnet-v1',
                                  'inception-resnet-v2'),
                      help = 'the cnn to use')
  parser$add_argument('--data-dir', type='character', help='the input data directory')
  parser$add_argument('--gpus', type='character',
                      help='the gpus will be used, e.g "0,1,2,3"')
  parser$add_argument('--batch-size', type='integer', default=128,
                      help='the batch size')
  parser$add_argument('--lr', type='double', default=.01,
                      help='the initial learning rate')
  parser$add_argument('--lr-factor', type='double', default=1,
                      help='times the lr with a factor for every lr-factor-epoch epoch')
  parser$add_argument('--lr-factor-epoch', type='double', default=1,
                      help='the number of epoch to factor the lr, could be .5')
  parser$add_argument('--lr-multifactor', type='character', 
                      help='the epoch at which the lr is changed, e.g "15,30,45"')
  parser$add_argument('--mom', type='double', default=.9,
                      help='momentum for sgd')
  parser$add_argument('--wd', type='double', default=.0001,
                      help='weight decay for sgd')
  parser$add_argument('--clip-gradient', type='double', default=5,
                      help='clip min/max gradient to prevent extreme value')
  parser$add_argument('--model-prefix', type='character',
                      help='the prefix of the model to load/save')
  parser$add_argument('--load-epoch', type='integer',
                      help="load the model on an epoch using the model-prefix")
  parser$add_argument('--num-round', type='integer', default=10,
                      help='the number of iterations over training data to train the model')
  parser$add_argument('--kv-store', type='character', default='local',
                      help='the kvstore type')
  parser$add_argument('--num-examples', type='integer', default=1281167,
                      help='the number of training examples')
  parser$add_argument('--num-classes', type='integer', default=1000,
                      help='the number of classes')
  parser$add_argument('--log-file', type='character', 
                      help='the name of log file')
  parser$add_argument('--log-dir', type='character', default="/tmp/",
                      help='directory of the log file')
  parser$add_argument('--train-dataset', type='character', default="train.rec",
                      help='train dataset name')
  parser$add_argument('--val-dataset', type='character', default="val.rec",
                      help="validation dataset name")
  parser$add_argument('--data-shape', type='integer', default=224,
                      help='set images shape')
  parser$add_argument('--depth', type='integer',
                      help='the depth for resnet, it can be a value among 18, 50, 101, 152, 200, 269')
  parser$parse_args()
}
args <- parse_args()

# network
if (args$network == 'inception-bn'){
  source("symbol_inception-bn.R")
} else if (args$network == 'googlenet'){
  if(args$data_shape < 299) stop(paste0("The data shape for ", args$network, " has to be at least 299"))
  source("symbol_googlenet.R")
} else if (args$network == 'inception-resnet-v1'){
  if(args$data_shape < 299) stop(paste0("The data shape for ", args$network, " has to be at least 299"))
  source("symbol_inception-resnet-v1.R")
}  else if (args$network == 'inception-resnet-v2'){
  if(args$data_shape < 299) stop(paste0("The data shape for ", args$network, " has to be at least 299"))
  source("symbol_inception-resnet-v2.R")
} else if (args$network == 'resnet'){
  source("symbol_resnet-v2.R")
} else{
  stop("Wrong network")
}
if (is.null(args$depth)){
  net <- get_symbol(args$num_classes)
} else{
  net <- get_symbol(args$num_classes, args$depth)
}

# train
source("train_model.R")
train_model.fit(args, net, get_iterator(args))


